### PR TITLE
TE-NS ubuntu migration 22.04

### PR DESCRIPTION
**Pull request by priyadarshitathagat**
Signed-off-by: priyadarshitathagat <tpriyadarshi@vmware.com>

### CODE REVIEW

**Comment by priyadarshitathagat** [comment]
The comments will be added post testing and validation. Would add comments as to why each change is required.

File: te_dp/Makefile [hunk 1]
```diff
@@ -39,20 +39,20 @@ BINDIR = bin
...
 
 CFLAGS     = -g -O0 -Wall -I.
-STAT_FLAGS = -pthread -lzmq -ljson
+STAT_FLAGS = -pthread -lzmq -ljson-c
 
 all: dp stat
 
 dp: $(DP_OBJ)
-	@$(CC) $(CFLAGS) -o $(DP_BIN) $^ $(DP_FLAGS)
+	@$(CC) $(CFLAGS) -o $(DP_BIN) $^ $(DP_FLAGS) -no-pie
```
Comment (aravindhank11) [inline te_dp/Makefile:55]: Could you please share the reasoning and need for `-no-pie` flag?

File: te/tedp_docker/Dockerfile [hunk 2]
```diff
@@ -53,7 +53,7 @@ RUN apt update && \
 
 # python related requirements
 COPY te/tedp_docker/requirements.txt /tmp/
-RUN pip3 install --upgrade "pip < 21.0"
+RUN pip3 install --upgrade "pip == 22.3.1"
```
Comment (aravindhank11) [inline te/tedp_docker/Dockerfile:56]: Should we pin pip to a version? Again this was done to avoid package breakage in pip specific to ubuntu 16 getting EOL. Is it needed?

File: te/te_docker/setup_postgres.sh [hunk 3]
```diff
@@ -32,7 +32,7 @@
 apt-get purge -y postgresql
 apt-get install -y wget
 wget --no-check-certificate --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
-sh -c 'echo "deb http://apt.postgresql.org/pub/repos/apt/ xenial-pgdg main" > /etc/apt/sources.list.d/PostgreSQL.list'
+sh -c 'echo "deb http://apt.postgresql.org/pub/repos/apt/ jammy-pgdg main" > /etc/apt/sources.list.d/PostgreSQL.list'
```
Comment (aravindhank11) [inline te/te_docker/setup_postgres.sh:35]: We might not need to get postgres from custom postgres repo. We could get it from ubuntu repo. Could we please consider removing this ?

Comment (aravindhank11) [reply te/te_docker/setup_postgres.sh:35]: Also move postgres to the latest version available. IG Ubuntu support psql 14

File: te/te_docker/Dockerfile [hunk 4]
```diff
@@ -75,12 +75,14 @@ missingok \n\
...
 
 #Creating a cron (Running cron job every 2 mins -- logrotate)
+RUN apt-get install -y cron
 RUN echo "*/2 * * * * /usr/sbin/logrotate /etc/logrotate.d/te-logs" > /etc/cron.d/cron_logrotate_te_logs
 RUN chmod 0644 /etc/cron.d/cron_logrotate_te_logs
 RUN crontab /etc/cron.d/cron_logrotate_te_logs
 
 COPY te_docker/requirements.txt /tmp/
-RUN pip3 install --upgrade "pip < 21.0"
+RUN apt install libpq-dev -y
```
Comment (aravindhank11) [inline te/te_docker/Dockerfile:84]: Why is dev package needed in the image?

File: te/TE.py [hunk 5]
```diff
@@ -1333,14 +1353,17 @@ def clean_api(self, jsonContent):
...
+                    tedp_host_config.append(HostConfig(user=te_dp_hosts[host_ip]['user']))
+                
             self.__setup_completed_tedps = self.__setup_completed_tedps.difference(te_dp_in_no_access)
             self.__connect_completed_tedps = self.__connect_completed_tedps.difference(te_dp_in_no_access)
 
             if(not(bool(te_dp_hosts))):
                 return self.__failure("No connected te dps to disconnect")
 
-            client = ParallelSSHClient(te_dp_hosts.keys(), host_config=te_dp_hosts, timeout = 240)
+            client = ParallelSSHClient(list(te_dp_hosts.keys()), host_config=tedp_host_config, timeout = 240)
```
Comment (aravindhank11) [inline te/TE.py:1366]: Is this change relevant to OS upgrade? If not, please consider raising a different one. 

If it is, please share details onto what is this change needed?

**Comment by priyadarshitathagat** [comment]
@aravindhank11 - we are in the process of testing and we are still in the process of fixing a lot of package dependency and modules changes that has come in due to python3.10.  I have compiled all the notes regarding each and every change, will update the PR with my comments soon as mentioned earlier. For now the PR is not open for review. 

Will tag everyone once all the changes as completed and open for review - I will have to minimise and trim a lot of changes that were added initially to get the compilation and docker going for testing.

File: te/TE.py (hunk 5 above)
Comment (priyadarshitathagat) [reply te/TE.py:1366]: Yes, these are all related to 22.04 migration - since we intend on using python3.10 and a lot of modules/api has changed.

File: te/tedp_docker/Dockerfile (hunk 2 above)
Comment (priyadarshitathagat) [reply te/tedp_docker/Dockerfile:56]: ACK.

File: te/tedp_docker/tedp_bin_builder.dockerfile [hunk 6]
```diff
@@ -50,12 +50,12 @@ RUN chmod 755 /tmp/setup.sh
...
 
 # uninstall libssl1.0 if any
-RUN apt remove --purge -y libssl-dev libssl-doc libssl1.0.0 openssl
+RUN apt remove --purge -y libssl-dev libssl-doc openssl
 
 RUN mkdir -pv $WORKDR/bin && mkdir $WORKDR/obj
 ADD te_dp/Makefile $WORKDR
 ADD te_dp/src $WORKDR/src
-RUN cd $WORKDR && make all
+RUN cd $WORKDR && make clean && make all
```
[the remark this answers is not in the document]
Comment (priyadarshitathagat) [reply te/tedp_docker/tedp_bin_builder.dockerfile:58]: ack - this was during my compilation - will remove

File: te_dp/Makefile (hunk 1 above)
Comment (priyadarshitathagat) [reply te_dp/Makefile:55]: https://stackoverflow.com/questions/73275183/gcc-assembly-x86-relocation-warning-creating-dt-textrel-in-a-pie 
it was due to this issue

File: te/te_docker/Dockerfile (hunk 4 above)
Comment (priyadarshitathagat) [reply te/te_docker/Dockerfile:84]: https://stackoverflow.com/questions/35104097/how-to-install-psycopg2-with-pg-config-error

File: te/tedp_docker/setup.sh [hunk 7]
```diff
@@ -70,16 +71,16 @@ cd /tmp/
 rm -rf openssl*
 
 #LIBCURL
-cd curl-7.67.0
```
[the remark this answers is not in the document]
Comment (priyadarshitathagat) [reply te/tedp_docker/setup.sh:73]: I will check with @vipinpr5991 also one and get back, while using curl-7.67 I was hitting an issue during compilation 
while i also tried curl-7.87 - but the the libcurl api has changed and stopped supporting a few, it would require a lot of C code changes to get this done. The intent was minimum penetration/changes in C code.

**Comment by priyadarshitathagat** [comment]
Created a branch to track every internal JIRA reg ubuntu migration bugs/fixes - 
Branch -> `tens-ubuntu-22.04-migration` 

Will consider addressing all the comments from here. Closing this one.
